pkg/cloud/azure/azure_storage.go

```diff
@@ -125,13 +130,21 @@ func parseAzureURL(
 		return conf, errors.Errorf("azure uri missing %q parameter", AzureAccountNameParam)
 	}
 
-	hasKeyCreds := conf.AzureConfig.AccountKey != ""
+	// Validate the authentication parameters are set correctly.
+	switch conf.AzureConfig.Auth {
+	case "", cloud.AuthParamSpecified:
+		hasKeyCreds := conf.AzureConfig.AccountKey != ""
 
-	hasRoleCreds := conf.AzureConfig.TenantID != "" && conf.AzureConfig.ClientID != "" && conf.AzureConfig.ClientSecret != ""
-	noRoleCreds := conf.AzureConfig.TenantID == "" && conf.AzureConfig.ClientID == "" && conf.AzureConfig.ClientSecret == ""
+		hasRoleCreds := conf.AzureConfig.TenantID != "" && conf.AzureConfig.ClientID != "" && conf.AzureConfig.ClientSecret != ""
```

Comment:
one idea to make these switch statements easier to read: declare a new " enum" type  in the azure conf that defines the 3 ways a user can authenticate to azure, i.e.:

```
type AzureAuth int32

const (
Implicit AzureAuth iota
LegacyKey 
RoleBased
)
```

then, you can use a single level switch statement to validate that all these params have been set correctly, as opposed to using a switches with conditionals.

Reply:
I like it!